protocol/claim_monero.go

```diff
@@ -0,0 +1,115 @@
+package protocol
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/athanorlabs/atomic-swap/coins"
+	"github.com/athanorlabs/atomic-swap/common"
+	"github.com/athanorlabs/atomic-swap/common/types"
+	mcrypto "github.com/athanorlabs/atomic-swap/crypto/monero"
+	"github.com/athanorlabs/atomic-swap/monero"
+
+	logging "github.com/ipfs/go-log"
+)
+
+var (
+	log = logging.Logger("protocol")
+)
+
+// GetClaimKeypair returns the private key pair required for a monero claim.
+// The key pair is the summation of each party's private spend and view keys:
+// (s_a + s_b) and (v_a + v_b).
+func GetClaimKeypair(
+	skA, skB *mcrypto.PrivateSpendKey,
+	vkA, vkB *mcrypto.PrivateViewKey,
+) *mcrypto.PrivateKeyPair {
+	skAB := mcrypto.SumPrivateSpendKeys(skA, skB)
+	vkAB := mcrypto.SumPrivateViewKeys(vkA, vkB)
+	kpAB := mcrypto.NewPrivateKeyPair(skAB, vkAB)
+	return kpAB
+}
+
+// ClaimMonero claims the XMR located in the wallet controlled by the private keypair `kpAB`.
+// If transferBack is true, it sweeps the XMR to `depositAddr`.
+func ClaimMonero(
+	ctx context.Context,
+	env common.Environment,
+	id types.Hash,
+	xmrClient monero.WalletClient,
+	walletScanHeight uint64,
+	kpAB *mcrypto.PrivateKeyPair,
+	depositAddr mcrypto.Address,
+	transferBack bool,
+) (mcrypto.Address, error) {
+	abAddr := kpAB.PublicKeyPair().Address(env)
+
+	conf := xmrClient.CreateWalletConf(fmt.Sprintf("swap-wallet-claim-%s", id))
+	abWalletCli, err := monero.CreateSpendWalletFromKeys(conf, kpAB, walletScanHeight)
+	if err != nil {
+		return "", err
+	}
+
+	if transferBack {
+		defer abWalletCli.CloseAndRemoveWallet()
+	} else {
+		abWalletCli.Close()
+		log.Infof("monero claimed in account %s with wallet file %s", abAddr, conf.WalletFilePath)
+		return abAddr, nil
+	}
+
+	log.Infof("monero claimed in account %s; transferring to deposit account %s",
+		abAddr, depositAddr)
+
+	err = mcrypto.ValidateAddress(string(depositAddr), env)
+	if err != nil {
+		log.Errorf(
+			"failed to transfer XMR out of swap wallet, dest address %s is invalid: %s",
+			abAddr,
+			err,
+		)
+		return "", err
+	}
+
+	err = waitUntilBalanceUnlocks(ctx, abWalletCli)
```

Comment:
This call isn't actually needed.  Waiting for the blocks to unlock is handled by the `SweepAll` method.

Reply:
`SweepAll` wait only 2 blocks but this waits for the balance to fully unlock (10 blocks). I think waiting just the 2 blocks is fine, should I remove this func?

Reply:
Sweep all waits for as many blocks as needed at the start if the SweepAll, but if new funds were received while waiting for those blocks, it won't continue waiting for the additional blocks.  I think that is good enough. I was trying to put all logic that would be used every time that `SweepAll` is called directly into our custom `SweepAll` method.